### PR TITLE
feat: skip levels

### PR DESCRIPTION
**Pull request by MichalBortkiewicz**

### CODE REVIEW

File: experiment_code/hackrl/experiment.py [hunk 1]
```diff
@@ -234,7 +290,7 @@ def make_ttyrec_envpool(threadpool, dataset_name, flags):
         aa_path = "/nle/nld-aa/nle_data"
         db.create(dbfilename)
         populate_db.add_nledata_directory(aa_path, "autoascend", dbfilename)
-        populate_db.add_altorg_directory(alt_path, "altorg", dbfilename)
+        # populate_db.add_altorg_directory(alt_path, "altorg", dbfilename)
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:293]: uncomment this

File: experiment_code/hackrl/experiment.py [hunk 2]
```diff
@@ -139,6 +140,61 @@ def _iter():
...
+                            else:
+                                return None
+
+                        # Vectorize the filtering function
+                        filter_vectorized = np.vectorize(filter_element, otypes=[object])
+
+                        # Apply the filtering function to the frame
+                        filtered_frame = filter_vectorized(frame)
+
+                        # Replace None values with element on the right or the left
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:180]: Since you are using a for loop you could just remember the previous dlvl and use its value to fill the Nones. It will be much simpler. You cannot go to the next level with a popup window.

File: experiment_code/hackrl/experiment.py [hunk 3]
```diff
@@ -46,6 +46,7 @@
 
 class TtyrecEnvPool:
     def __init__(self, flags, dataset_name, dataset_scores, **dataset_kwargs):
+        self.flags = flags
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:49]: Maybe its better just to pass this argument omitted_dlvls?
`self.omitted_dlvls = omitted_dlvls`

File: experiment_code/hackrl/experiment.py [hunk 4]
```diff
@@ -139,6 +140,61 @@ def _iter():
...
+                                    ):
+                                        filtered_frame[i, j] = filtered_frame[i, j + 1]
+                                    elif j - 1 >= 0 and filtered_frame[i, j - 1] is not None:
+                                        filtered_frame[i, j] = filtered_frame[i, j - 1]
+
+                        return filtered_frame
+
+                    dlvl = _filter_2d_dlvl_array(dlvl)
+
+                    mask = torch.from_numpy(dlvl > self.flags.omitted_dlvls) if self.flags.omitted_dlvls > 0 else torch.ones_like(mb_tensors["timestamps"]).bool()
```
Comment (BartekCupial) [inline experiment_code/hackrl/experiment.py:196]: You don't use the mask in `final_mb`

File: experiment_code/hackrl/experiment.py (hunk 2 above)
Comment (BartekCupial) [reply experiment_code/hackrl/experiment.py:180]:
```
                        # Replace None values with dlvl from previous frame
                        for i in range(filtered_frame.shape[0]):
                            prev_dlvl = 1
                            for j in range(filtered_frame.shape[1]):
                                if filtered_frame[i, j] is None:
                                    filtered_frame[i, j] = prev_dlvl
                                prev_dlvl = filtered_frame[i, j]
```

Comment (BartekCupial) [reply experiment_code/hackrl/experiment.py:180]: Actually, the code above is also wrong. It doesn't handle the case when `filtered_frame[i, 0] is None and filtered_frame[i, j] > 1`. Your version also would be able to handle it. It's possible to pass the prev_dlvl as a state, but I think it's better to just use the future or past dlvl. So your idea is fine with a modification to look further (it's possible that there will be more Nones than 2.

Comment (BartekCupial) [reply experiment_code/hackrl/experiment.py]: Actually, never mind. prev_dlvl as a state is easy to implement and I've done it. You can see a new commit.
